java/ql/src/experimental/semmle/code/java/security/DecompressionBomb.qll

```diff
@@ -0,0 +1,390 @@
+      }
+    }
+
+    /**
+     * A type that is responsible for `CompressorStreamFactory` Class
+     */
+    class TypeCompressors extends RefType {
+      TypeCompressors() {
+        this.getASupertype*()
+            .hasQualifiedName("org.apache.commons.compress.compressors", "CompressorStreamFactory")
+      }
+    }
+
+    /**
+     * Gets `n1` and `n2` which `CompressorInputStream n2 = new CompressorStreamFactory().createCompressorInputStream(n1)`
+     * or `ArchiveInputStream n2 = new ArchiveStreamFactory().createArchiveInputStream(n1)` or
+     * `n1.read(n2)`,
+     * second one is added because of sanitizer, we want to compare return value of each `read` or similar method
+     * that whether there is a flow to a comparison between total read of decompressed stream and a constant value
+     */
+    private class CompressorsAndArchiversAdditionalTaintStep extends DecompressionBomb::AdditionalStep
+    {
+      override predicate step(DataFlow::Node n1, DataFlow::Node n2) {
+        exists(Call call |
+          // Constructors
+          (
+            call.getCallee().getDeclaringType() instanceof TypeCompressors
+            or
+            call.getCallee().getDeclaringType() instanceof TypeArchivers
+          ) and
+          call.getArgument(0) = n1.asExpr() and
+          call = n2.asExpr()
+        )
+      }
+    }
+
+    /**
+     * The methods that read bytes and belong to `CompressorInputStream` or `ArchiveInputStream` Types
+     */
+    class ReadInputStreamCall extends DecompressionBomb::BombReadInputStreamCall {
+      ReadInputStreamCall() {
+        (
+          this.getReceiverType() instanceof TypeArchiveInputStream
+          or
+          this.getReceiverType() instanceof TypeCompressorInputStream
+        ) and
+        this.getCallee().hasName(["read", "readNBytes", "readAllBytes"])
+      }
+    }
+  }
+}
+
+/**
+ * Providing Decompression sinks and additional taint steps for `net.lingala.zip4j.io` package
+ */
+module Zip4j {
+  /**
+   * A type that is responsible for `ZipInputStream` Class
+   */
+  class TypeZipInputStream extends RefType {
+    TypeZipInputStream() {
+      this.hasQualifiedName("net.lingala.zip4j.io.inputstream", "ZipInputStream")
+    }
+  }
+
+  /**
+   * The methods that read bytes and belong to `CompressorInputStream` or `ArchiveInputStream` Types
+   */
+  class ReadInputStreamCall extends DecompressionBomb::BombReadInputStreamCall {
+    ReadInputStreamCall() {
+      this.getReceiverType() instanceof TypeZipInputStream and
+      this.getMethod().hasName(["read", "readNBytes", "readAllBytes"])
+    }
+  }
+
+  class Sink extends DecompressionBomb::Sink {
+    Sink() {
+      this.asExpr() = any(ReadInputStreamCall r).getQualifier()
+      or
+      exists(ConstructorCall call | call.getConstructedType() instanceof TypeZipInputStream |
+        this.asExpr() = call.getArgument(0)
+      )
+    }
+  }
+}
```

Comment:
```suggestion
```
The first part of this is covered by `ReadInputStreamQualifierSink`. The second part I think should be deleted - for all the other libraries you have deleted the sinks relating to just constructing an input stream. (The test results will have to be updated.)